### ISSUE TITLE
100% keytrack on a comb filter is out of tune

### ISSUE THREAD

**Issue by unfa** [surge-synthesizer/surge#3248]
**Bug Description:**
After setting a filter keytracking to 100% I'd expect eh filter to play in unison with the oscillators.
Even after manually tunining the base filter frequency to match the osciallator, the octave of keytrack is not an octave of the oscillator.

**Surge Version**
This information is found on the About screen, which you get to from the bottom right menu

* Version: 1.7.1
* Plugin Type: VST3
* Bitness: 64-bit

**Reproduction Steps:**

1. Disable the Oscillator 1
1. Enable noise source and turn it down to -50 dB
2. Enable Comb filer and turn the resonance to 100%
3. Set Filter 1 keytrack to 100%
4. Set filter base frequency to 440 Hz
5. Set the keytrack reference note to A4
6. Play a Cm chord and tell me if it's in tune

**Expected Behavior:**
I'd expect keytrack of 100% to produce same intervals as the oscillators, but it's not the case.

**Screenshots:**

An in-depth demonstration of the problem and what I'm trying to achieve.
https://youtu.be/gme759i2-KQ

**Computer Information (please complete the following information):**
 - OS: Manjaro Linux
 - Host: Ardour 6.5

**Additional information:**

I haven't tested the keytrack intervals with other types of filters, but the comb filter should the easiest to play in tune, since there's no "pole tuning" involved for like it is with the resonance peaks of lowpass filters etc.

**Comment by baconpaul** [comment]
So I'm not 100% sure this is a fixable bug but it might be. Before I dig to much lets make sure we see the same thing.

[fa.zip](https://github.com/surge-synthesizer/surge/files/5605045/fa.zip)

Here's a patch which shows kinda what I think you are trying to do

If you play a C-Minor chord low in the keyboard (C2/Eb2/G2) it is perfectly in tune. If you play high on the keyboard (C6etc...) it is not. At least at 44.1k hz. If you increase to 96k it gets better.

This is because the comb filter works with an FIR onto a delay line. And those high pitches have increasingly short delay lines compared to the window available.  If you increase the sample rate the in-tunedness decreases also. (Filter resonance is one place where you inevitably have some sample rate dependency and we generally have that across most of our feedback type circuits, of which comb is one).

There's a suspicious -1 I see in the filter which I want to test before I say 100% that this is "not a bug" - and an off by one error would indeed cause frequency dependent mis-tuning. So I'll dig. Just want to be sure we are digging on the same thing!

**Comment by baconpaul** [comment]
Hmm yeah right that -1 was suspicious.

If I apply this diff

```
diff --git a/src/common/dsp/FilterCoefficientMaker.cpp b/src/common/dsp/FilterCoefficientMaker.cpp
index a47005d2..e0798baf 100644
--- a/src/common/dsp/FilterCoefficientMaker.cpp
+++ b/src/common/dsp/FilterCoefficientMaker.cpp
@@ -434,9 +434,9 @@ void FilterCoefficientMaker::Coeff_LP4L(float freq, float reso, int subtype)
 
 void FilterCoefficientMaker::Coeff_COMB(float freq, float reso, int subtype)
 {
-   float dtime = (1.f / 440.f) * storage->note_to_pitch_ignoring_tuning(-freq);
-   dtime = dtime * dsamplerate_os -
-           FIRoffset; // 1 sample for feedback, 1 sample for the IIR-filter without resonance
+   float dtime = (1.f / 440.f) * storage->note_to_pitch_inv_ignoring_tuning(freq);
+   dtime = dtime * dsamplerate_os; // -
+           //FIRoffset; // 1 sample for feedback, 1 sample for the IIR-filter without resonance
```

the comb filter is exactly in tune with that patch across the whole keyboard.

Let me ponder the consequences of that on a less thanksgiving-y day. If you can just confirm that the patch replicates the problem you see also I'll figure out why that FIRoffset is there (or why it is re-compensated for in the filter SSE code). But that's not a quick-fix-before-turkey type thing. 

Thanks as always for all your support of surge.

**Comment by unfa** [comment]
Thanks for looking into this.
So have you managed to get it in tune? Did I understand you correctly?
And it needs testing to see if that doesn't break something else?

Honestly I don't know how to load the file you've sent me - sorry.
Don't hurry though, I can wait :D

**Comment by baconpaul** [comment]
If you unzip that file you get an .fxp. Put it in your documents/surge folder (whose name is on the about screen) in a directory called "TEST" and restart surge. Use the patch browser and you will see TEST as user patches with that fxp 

And yes with that diff applied to the C++ it is 100% in tune.

@mkruselj were discussing what to do about that fact and compatibility with old patches....

**Comment by mkruselj** [comment]
My thinking is: this change is coming too late in this release cycle, and it would affect how projects sound for people out there. Hence I would not rush it.

In Surge 1.9 we plan on expanding the parameter set, so filters could (and will!) get more sliders than just cutoff and resonance. For comb filter, one of these sliders can replace all the current comb filter subtypes (50% wet positive, 100% wet positive, 50% wet negative, 100% wet negative), by being the Mix slider with range of -100% to 100%. We would convert the old patches on load so that the subtypes are now the slider value (-100%, -50%, 50%, 100%). And then we introduce the fixed comb filter as subtype 1, while subtype 2 is the legacy comb filter.

Precedent for this is the notch filter in Surge, whose "Mild" mode used to be called Legacy (it had smaller resonance range initially, Claes fixed it later and added the fix as a new subtype which then became the default notch subtype).

**Comment by baconpaul** [comment]
@unfa we chatted in our discord and have a plan to fix this for 18 and fix more robustly for the subsequent 19 release

**Comment by unfa** [comment]
Thanks! I'm excited to do some physical modelling in Surge! Should the other filters tune properly as well?  I guess I should test them all once it's in.
Also: maybe the default filter cutoff position could be so just turning up the keytrack to 100% will make the filter be in tune with the default sawtooth oscillator?

**Comment by mkruselj** [comment]
Other filters tune properly.

**Comment by unfa** [comment]
Aha! That's great as I can make a pipe organ sound :D
BTW - in Vital the comb filter has two additional controls that seem to be controlling an additional low/high shelve pair (or a bandpass?) - there's width and frequency. This allows for some amazing sounds, because we can control how fast the harmonics decay. If you're extending the filters, I guess that'd be an amazing addition.

**Comment by mkruselj** [comment]
Not sure if we just want to outright copy features from every other synth, though... Besides you can kinda rig that already in Surge with serial filter routing (or if you've already used up two filters in a scene, route scene A to scene B using audio in oscillator then you get 2 more filters to play with 🙂)

**Comment by baconpaul** [comment]
Fix inbound which will ship in 1.8 and be in the nightlies today. The filter subtype now has a menu for "correctly tune comb filter" which defaults to true for new patches and false for streamed old ones etc.
